cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java

```diff
@@ -0,0 +1,273 @@
+              run, datasetInstance, accessType, stopTime, component);
+    accessRegistryTable.put(getDatasetKey(datasetInstance, run, accessType, component),
+                          STOP_COL_BYTE, Bytes.toBytes(stopTime));
+    accessRegistryTable.put(getProgramKey(run, datasetInstance, accessType, component),
+                          STOP_COL_BYTE, Bytes.toBytes(stopTime));
+  }
+
+  public void addAccess(Id.Run run, Id.DatasetInstance datasetInstance,
+                        AccessType accessType, String metadata, @Nullable Id component) {
+    LOG.trace("Recording access run={}, dataset={}, accessType={}, metadata={}, component={}",
+              run, datasetInstance, accessType, metadata, component);
+    accessRegistryTable.put(getDatasetKey(datasetInstance, run, accessType, component),
+                          METADATA_COLS_BYTE, Bytes.toBytes(metadata));
+    accessRegistryTable.put(getProgramKey(run, datasetInstance, accessType, component),
+                          METADATA_COLS_BYTE, Bytes.toBytes(metadata));
+  }
+
+  public Set<Relation> getRelations(Id.DatasetInstance datasetInstance, long start, long end) {
+    return scanRelations(getDatasetScanStartKey(datasetInstance, end),
+                         getDatasetScanEndKey(datasetInstance),
+                         start);
+  }
+
+  public Set<Relation> getRelations(Id.Program program, long start, long end) {
+    return scanRelations(getProgramScanStartKey(program, end),
+                         getProgramScanEndKey(program),
+                         start);
+  }
+
+  private Set<Relation> scanRelations(byte[] startKey, byte[] endKey, long startTime) {
+    Set<Relation> relations = new HashSet<>();
+    Scanner scanner = accessRegistryTable.scan(startKey, endKey);
+    try {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
+        }
+        Long stopTime = row.getLong(STOP_COL_BYTE);
+        // TODO: convert this check into a scan filter for performance reasons
+        if (stopTime == null || stopTime > startTime) {
+          Relation relation = toRelation(row);
+          relations.add(relation);
+        }
+      }
+    } finally {
+      scanner.close();
+    }
+    return relations;
+  }
+
+  private static byte[] getDatasetKey(Id.DatasetInstance datasetInstance, Id.Run run,
+                                      AccessType accessType, @Nullable Id component) {
+    long invertedStartTime = getInvertedStartTime(run);
+    MDSKey.Builder builder = new MDSKey.Builder();
+    addDataset(builder, datasetInstance);
+    builder.add(invertedStartTime);
+    addProgram(builder, run.getProgram());
+    builder.add(run.getId());
+    builder.add(accessType.getType());
+    addComponent(builder, component);
+
+    return builder.build().getKey();
+  }
+
+  private static byte[] getProgramKey(Id.Run run, Id.DatasetInstance datasetInstance,
```

Comment:
better not to use `private static` but just `private` (static is generally evil).